### ISSUE TITLE
7.1.0rc1 crashes with `Unrecoverable error while evaluating node REPOSITORY_DIRECTORY`

### ISSUE THREAD

**Issue by criemen** [bazelbuild/bazel#21483]
### Description of the bug:

Buildling our project (working fine with 7.0.2) on Ubuntu 22.04 crashes with

```
FATAL: bazel crashed due to an internal error. Printing stack trace:
java.lang.RuntimeException: Unrecoverable error while evaluating node 'REPOSITORY_DIRECTORY:@@rules_nodejs~~node~nodejs' (requested by nodes 'PACKAGE_LOOKUP:@@rules_nodejs~~node~nodejs//')
        at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:550)
        at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:414)
        at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
Caused by: java.lang.NullPointerException
        at com.google.common.base.Preconditions.checkNotNull(Preconditions.java:903)
        at com.google.devtools.build.lib.rules.repository.RepositoryDelegatorFunction.fetchRepository(RepositoryDelegatorFunction.java:439)
        at com.google.devtools.build.lib.rules.repository.RepositoryDelegatorFunction.compute(RepositoryDelegatorFunction.java:205)
        at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:461)
        ... 7 more
BAZEL FAILED
```

The error is reproducible. 

### Which category does this issue belong to?

Core

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

I've not worked on getting a small reproducing example yet (our repo is unfortunately closed source).
This looks somewhat similar other crashes, though, so I wonder if this is necessary right now.

### Which operating system are you running Bazel on?

Ubuntu 22.04

### What is the output of `bazel info release`?

release 7.1.0rc1

### What's the output of `git remote get-url origin; git rev-parse HEAD` ?

private repository

**Comment by meisterT** [comment]
This was last touched in https://github.com/bazelbuild/bazel/pull/21182/files

**Comment by meteorcloudy** [comment]
@criemen Is it possible to construct a minimal reproducible case? It'll make it much easier to diagnose this problem.

**Comment by meteorcloudy** [comment]
`bazel --bisect=7.0.0..release-7.1.0 build --nobuild //:typescript` gives 18ba4492c752530b168daa563987400f7f9df4bb

**Comment by Wyverald** [comment]
Same bug on master, culprit is https://github.com/bazelbuild/bazel/commit/18ba4492c752530b168daa563987400f7f9df4bb. Investigating.

**Comment by Wyverald** [comment]
This part of the rules_nodejs code triggers the bug: https://github.com/bazelbuild/rules_nodejs/blob/943800d8d601872dae53f5c1e17414d316834cac/nodejs/private/nodejs_repo_host_os_alias.bzl#L34-L38

This is because we try to watch the symlink target, but the target in this case actually points into where another repo (in this case `@@nodejs_darwin_arm64`) would be fetched. The problem is that that repo isn't actually defined (IOW, it doesn't exist); in the Bzlmod world, that repo's name would actually be `@@rules_nodejs~~nodejs~nodejs_darwin_arm64`.

I think there are two things that need to be fixed:
1. The parameter `rctx.symlink(watch_target=)` shouldn't exist. The symlink target's existence and its contents should not affect the creation of the symlink at all, so there's no reason to ever watch it.
2. Watching a path that corresponds to a nonexistent/undefined repo should either result in an explicit error, or it should be the same as watching any nonexistent file. An explicit error makes sense because this file will technically never exist (all bets are off if you go creating things in `$outputBase/external` manually...).

**Comment by Wyverald** [comment]
Number 1 is P1 (that should solve the immediate bug); number 2 is P2 or P3 since nobody should explicitly call `rctx.watch()` on such a path.

**Comment by criemen** [comment]
> The problem is that that repo isn't actually defined (IOW, it doesn't exist); in the Bzlmod world, that repo's name would actually be @@rules_nodejs~~nodejs~nodejs_darwin_arm64.

Would you mind filing a `rules_nodejs` bug for refering to the wrong repository with `bzlmod` enabled?

**Comment by Wyverald** [comment]
re point 2: I ended up making it the same as watching any other nonexistent path (i.e. _not_ throwing) because that repo can technically become defined at some point (see unit test in #21562).

> > The problem is that that repo isn't actually defined (IOW, it doesn't exist); in the Bzlmod world, that repo's name would actually be @@rules_nodejs~~nodejs~nodejs_darwin_arm64.
> 
> Would you mind filing a `rules_nodejs` bug for refering to the wrong repository with `bzlmod` enabled?

Done: https://github.com/bazelbuild/rules_nodejs/issues/3720

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 7.1.0 RC2](https://github.com/bazelbuild/bazel/releases/tag/7.1.0rc2). Please test out the release candidate and report any issues as soon as possible.
If you're using Bazelisk, you can point to the latest RC by setting USE_BAZEL_VERSION=7.1.0rc2. Thanks!
